[BACKEND] Dedup elementwise in LLVM IR based on constancy

aakhundov opened the pull request:
### Summary

When Triton GPU IR is lowered into LLVM IR, we can make use of the constancy information about the result of the elementwise ops to deduplicate otherwise redundant computation. That is the contribution of this PR: the constancy is checked and, if possible, some of the values in LLVM IR are reused multiple times instead of computing equal values separately.

The change is beneficial for the PyTorch 2 / TorchInductor-generated Triton code, as the leftmost sub-indices extracted from the flat index by div / mod operations can be equal, given sufficiently large 2^n factor in the rightmost rightmost dimension(s). This makes the computation resulting in those sub-indices redundant. Consequently, under the necessary constancy conditions, the redundant indexing arithmetics can be deduplicated. We observe up to 29% decrease in the latency of some of our jagged tensor kernels (see the "Benchmarking" section below).

### Example

For this simple Triton kernel:

```python
@triton.jit
def triton_kernel(in_ptr, out_ptr, xnumel, XBLOCK: tl.constexpr):
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[:]
    xmask = xindex < xnumel

    x1 = xindex // 256

    tmp = tl.load(in_ptr + x1, xmask)
    tl.store(out_ptr + xindex, tmp, xmask)


BLOCK_SIZE = 1024
input_t = torch.randn((1234,)).half().cuda()
output_t = torch.randn((1234, 256)).half().cuda()
grid = (triton.cdiv(output_t.numel(), BLOCK_SIZE),)
res = triton_kernel[grid](
    input_t,
    output_t,
    output_t.numel(),
    XBLOCK=BLOCK_SIZE,
    num_warps=4,
)
``` 

This is the Triton GPU IR annotated with the axis analysis info (including the constancy):

```
@triton_kernel_0d1d2de
%cst = arith.constant dense<256> : tensor<1024xi32, #triton_gpu.blocked<{sizePerThread = [8], threadsPerWarp = [32], warpsPerCTA = [4], order = [0], CTAsPerCGA = [1], CTASplitNum = [1], CTAOrder = [0]}>> => contiguity = [1], divisibility = [256], constancy = [1024], constant_value = 256
%c1024_i32 = arith.constant 1024 : i32 => contiguity = [1], divisibility = [1024], constancy = [1], constant_value = 1024
%0 = tt.get_program_id x : i32 => contiguity = [1], divisibility = [1], constancy = [1], constant_value = <none>
%1 = arith.muli %0, %c1024_i32 : i32 => contiguity = [1], divisibility = [1024], constancy = [1], constant_value = <none>
%2 = tt.make_range {end = 1024 : i32, start = 0 : i32} : tensor<1024xi32, #triton_gpu.blocked<{sizePerThread = [8], threadsPerWarp = [32], warpsPerCTA = [4], order = [0], CTAsPerCGA = [1], CTASplitNum = [1], CTAOrder = [0]}>> => contiguity = [1024], divisibility = [1073741824], constancy = [1], constant_value = <none>
%3 = tt.splat %1 : (i32) -> tensor<1024xi32, #triton_gpu.blocked<{sizePerThread = [8], threadsPerWarp = [32], warpsPerCTA = [4], order = [0], CTAsPerCGA = [1], CTASplitNum = [1], CTAOrder = [0]}>> => contiguity = [1], divisibility = [1024], constancy = [1024], constant_value = <none>
%4 = arith.addi %3, %2 : tensor<1024xi32, #triton_gpu.blocked<{sizePerThread = [8], threadsPerWarp = [32], warpsPerCTA = [4], order = [0], CTAsPerCGA = [1], CTASplitNum = [1], CTAOrder = [0]}>> => contiguity = [1024], divisibility = [1024], constancy = [1], constant_value = <none>
%5 = tt.splat %arg2 : (i32) -> tensor<1024xi32, #triton_gpu.blocked<{sizePerThread = [8], threadsPerWarp = [32], warpsPerCTA = [4], order = [0], CTAsPerCGA = [1], CTASplitNum = [1], CTAOrder = [0]}>> => contiguity = [1], divisibility = [16], constancy = [1024], constant_value = <none>
%6 = "triton_gpu.cmpi"(%4, %5) <{predicate = 2 : i64}> : (tensor<1024xi32, #triton_gpu.blocked<{sizePerThread = [8], threadsPerWarp = [32], warpsPerCTA = [4], order = [0], CTAsPerCGA = [1], CTASplitNum = [1], CTAOrder = [0]}>>, tensor<1024xi32, #triton_gpu.blocked<{sizePerThread = [8], threadsPerWarp = [32], warpsPerCTA = [4], order = [0], CTAsPerCGA = [1], CTASplitNum = [1], CTAOrder = [0]}>>) -> tensor<1024xi1, #triton_gpu.blocked<{sizePerThread = [8], threadsPerWarp = [32], warpsPerCTA = [4], order = [0], CTAsPerCGA = [1], CTASplitNum = [1], CTAOrder = [0]}>> => contiguity = [1], divisibility = [1], constancy = [16], constant_value = <none>
%7 = arith.divsi %4, %cst : tensor<1024xi32, #triton_gpu.blocked<{sizePerThread = [8], threadsPerWarp = [32], warpsPerCTA = [4], order = [0], CTAsPerCGA = [1], CTASplitNum = [1], CTAOrder = [0]}>> => contiguity = [1], divisibility = [1], constancy = [256], constant_value = <none>
%8 = tt.splat %arg0 : (!tt.ptr<f16, 1>) -> tensor<1024x!tt.ptr<f16, 1>, #triton_gpu.blocked<{sizePerThread = [8], threadsPerWarp = [32], warpsPerCTA = [4], order = [0], CTAsPerCGA = [1], CTASplitNum = [1], CTAOrder = [0]}>> => contiguity = [1], divisibility = [16], constancy = [1024], constant_value = <none>
%9 = tt.addptr %8, %7 : tensor<1024x!tt.ptr<f16, 1>, #triton_gpu.blocked<{sizePerThread = [8], threadsPerWarp = [32], warpsPerCTA = [4], order = [0], CTAsPerCGA = [1], CTASplitNum = [1], CTAOrder = [0]}>>, tensor<1024xi32, #triton_gpu.blocked<{sizePerThread = [8], threadsPerWarp = [32], warpsPerCTA = [4], order = [0], CTAsPerCGA = [1], CTASplitNum = [1], CTAOrder = [0]}>> => contiguity = [1], divisibility = [2], constancy = [256], constant_value = <none>
%10 = tt.load %9, %6 {cache = 1 : i32, evict = 1 : i32, isVolatile = false} : tensor<1024xf16, #triton_gpu.blocked<{sizePerThread = [8], threadsPerWarp = [32], warpsPerCTA = [4], order = [0], CTAsPerCGA = [1], CTASplitNum = [1], CTAOrder = [0]}>> => contiguity = [1], divisibility = [1], constancy = [16], constant_value = <none>
%11 = tt.splat %arg1 : (!tt.ptr<f16, 1>) -> tensor<1024x!tt.ptr<f16, 1>, #triton_gpu.blocked<{sizePerThread = [8], threadsPerWarp = [32], warpsPerCTA = [4], order = [0], CTAsPerCGA = [1], CTASplitNum = [1], CTAOrder = [0]}>> => contiguity = [1], divisibility = [16], constancy = [1024], constant_value = <none>
%12 = tt.addptr %11, %4 : tensor<1024x!tt.ptr<f16, 1>, #triton_gpu.blocked<{sizePerThread = [8], threadsPerWarp = [32], warpsPerCTA = [4], order = [0], CTAsPerCGA = [1], CTASplitNum = [1], CTAOrder = [0]}>>, tensor<1024xi32, #triton_gpu.blocked<{sizePerThread = [8], threadsPerWarp = [32], warpsPerCTA = [4], order = [0], CTAsPerCGA = [1], CTASplitNum = [1], CTAOrder = [0]}>> => contiguity = [1024], divisibility = [16], constancy = [1], constant_value = <none>
```

Before this PR, the LLVM IR (shown below) and the PTX (not shown here) both contain 8 redundant computations of the same `x1` value for each element in the thread; the 8 loads are performed from the 8 independent pointers containing the same address:

```
define void @triton_kernel_0d1d2de(ptr addrspace(1) %0, ptr addrspace(1) %1, i32 %2) local_unnamed_addr !dbg !5 {
  %4 = tail call i32 @llvm.nvvm.read.ptx.sreg.tid.x(), !dbg !8
  %5 = shl i32 %4, 3, !dbg !8
  %6 = and i32 %5, 1016, !dbg !8
  %7 = tail call i32 asm "mov.u32 $0, %ctaid.x;", "=r"() #1, !dbg !9
  %8 = shl i32 %7, 10, !dbg !10
  %9 = or i32 %8, %6, !dbg !11
  %10 = or i32 %9, 1, !dbg !11
  %11 = or i32 %9, 2, !dbg !11
  %12 = or i32 %9, 3, !dbg !11
  %13 = or i32 %9, 4, !dbg !11
  %14 = or i32 %9, 5, !dbg !11
  %15 = or i32 %9, 6, !dbg !11
  %16 = or i32 %9, 7, !dbg !11
  %17 = icmp slt i32 %9, %2, !dbg !12
  %18 = icmp slt i32 %10, %2, !dbg !12
  %19 = icmp slt i32 %11, %2, !dbg !12
  %20 = icmp slt i32 %12, %2, !dbg !12
  %21 = icmp slt i32 %13, %2, !dbg !12
  %22 = icmp slt i32 %14, %2, !dbg !12
  %23 = icmp slt i32 %15, %2, !dbg !12
  %24 = icmp slt i32 %16, %2, !dbg !12
  %25 = sdiv i32 %9, 256, !dbg !13
  %26 = sdiv i32 %10, 256, !dbg !13
  %27 = sdiv i32 %11, 256, !dbg !13
  %28 = sdiv i32 %12, 256, !dbg !13
  %29 = sdiv i32 %13, 256, !dbg !13
  %30 = sdiv i32 %14, 256, !dbg !13
  %31 = sdiv i32 %15, 256, !dbg !13
  %32 = sdiv i32 %16, 256, !dbg !13
  %33 = sext i32 %25 to i64, !dbg !14
  %34 = getelementptr half, ptr addrspace(1) %0, i64 %33, !dbg !14
  %35 = sext i32 %26 to i64, !dbg !14
  %36 = getelementptr half, ptr addrspace(1) %0, i64 %35, !dbg !14
  %37 = sext i32 %27 to i64, !dbg !14
  %38 = getelementptr half, ptr addrspace(1) %0, i64 %37, !dbg !14
  %39 = sext i32 %28 to i64, !dbg !14
  %40 = getelementptr half, ptr addrspace(1) %0, i64 %39, !dbg !14
  %41 = sext i32 %29 to i64, !dbg !14
  %42 = getelementptr half, ptr addrspace(1) %0, i64 %41, !dbg !14
  %43 = sext i32 %30 to i64, !dbg !14
  %44 = getelementptr half, ptr addrspace(1) %0, i64 %43, !dbg !14
  %45 = sext i32 %31 to i64, !dbg !14
  %46 = getelementptr half, ptr addrspace(1) %0, i64 %45, !dbg !14
  %47 = sext i32 %32 to i64, !dbg !14
  %48 = getelementptr half, ptr addrspace(1) %0, i64 %47, !dbg !14
  %49 = tail call i16 asm sideeffect "mov.u16 $0, 0x0;\0A\09@$2 ld.global.b16 { $0 }, [ $1 + 0 ];", "=c,l,b"(ptr addrspace(1) %34, i1 %17) #1, !dbg !15
  %50 = tail call i16 asm sideeffect "mov.u16 $0, 0x0;\0A\09@$2 ld.global.b16 { $0 }, [ $1 + 0 ];", "=c,l,b"(ptr addrspace(1) %36, i1 %18) #1, !dbg !15
  %51 = tail call i16 asm sideeffect "mov.u16 $0, 0x0;\0A\09@$2 ld.global.b16 { $0 }, [ $1 + 0 ];", "=c,l,b"(ptr addrspace(1) %38, i1 %19) #1, !dbg !15
  %52 = tail call i16 asm sideeffect "mov.u16 $0, 0x0;\0A\09@$2 ld.global.b16 { $0 }, [ $1 + 0 ];", "=c,l,b"(ptr addrspace(1) %40, i1 %20) #1, !dbg !15
  %53 = tail call i16 asm sideeffect "mov.u16 $0, 0x0;\0A\09@$2 ld.global.b16 { $0 }, [ $1 + 0 ];", "=c,l,b"(ptr addrspace(1) %42, i1 %21) #1, !dbg !15
  %54 = tail call i16 asm sideeffect "mov.u16 $0, 0x0;\0A\09@$2 ld.global.b16 { $0 }, [ $1 + 0 ];", "=c,l,b"(ptr addrspace(1) %44, i1 %22) #1, !dbg !15
  %55 = tail call i16 asm sideeffect "mov.u16 $0, 0x0;\0A\09@$2 ld.global.b16 { $0 }, [ $1 + 0 ];", "=c,l,b"(ptr addrspace(1) %46, i1 %23) #1, !dbg !15
  %56 = tail call i16 asm sideeffect "mov.u16 $0, 0x0;\0A\09@$2 ld.global.b16 { $0 }, [ $1 + 0 ];", "=c,l,b"(ptr addrspace(1) %48, i1 %24) #1, !dbg !15
  %57 = sext i32 %9 to i64, !dbg !16
  %58 = getelementptr half, ptr addrspace(1) %1, i64 %57, !dbg !16
  %59 = insertelement <2 x i16> undef, i16 %49, i64 0, !dbg !17
  %60 = insertelement <2 x i16> %59, i16 %50, i64 1, !dbg !17
  %61 = bitcast <2 x i16> %60 to i32, !dbg !17
  %62 = insertelement <2 x i16> undef, i16 %51, i64 0, !dbg !17
  %63 = insertelement <2 x i16> %62, i16 %52, i64 1, !dbg !17
  %64 = bitcast <2 x i16> %63 to i32, !dbg !17
  %65 = insertelement <2 x i16> undef, i16 %53, i64 0, !dbg !17
  %66 = insertelement <2 x i16> %65, i16 %54, i64 1, !dbg !17
  %67 = bitcast <2 x i16> %66 to i32, !dbg !17
  %68 = insertelement <2 x i16> undef, i16 %55, i64 0, !dbg !17
  %69 = insertelement <2 x i16> %68, i16 %56, i64 1, !dbg !17
  %70 = bitcast <2 x i16> %69 to i32, !dbg !17
  tail call void asm sideeffect "@$5 st.global.v4.b32 [ $4 + 0 ], { $0, $1, $2, $3 };", "r,r,r,r,l,b"(i32 %61, i32 %64, i32 %67, i32 %70, ptr addrspace(1) %58, i1 %17) #1, !dbg !17
  ret void, !dbg !18
}
```

After this PR, the redundant computations in the LLVM IR shown above are deduplicated; as a result, the 8 loads are performed from the same pointer:

```
define void @triton_kernel_0d1d2de(ptr addrspace(1) %0, ptr addrspace(1) %1, i32 %2) local_unnamed_addr !dbg !5 {
  %4 = tail call i32 @llvm.nvvm.read.ptx.sreg.tid.x(), !dbg !8
  %5 = shl i32 %4, 3, !dbg !8
  %6 = and i32 %5, 1016, !dbg !8
  %7 = tail call i32 asm "mov.u32 $0, %ctaid.x;", "=r"() #1, !dbg !9
  %8 = shl i32 %7, 10, !dbg !10
  %9 = or i32 %8, %6, !dbg !11
  %10 = icmp slt i32 %9, %2, !dbg !12
  %11 = sdiv i32 %9, 256, !dbg !13
  %12 = sext i32 %11 to i64, !dbg !14
  %13 = getelementptr half, ptr addrspace(1) %0, i64 %12, !dbg !14
  %14 = tail call i16 asm sideeffect "mov.u16 $0, 0x0;\0A\09@$2 ld.global.b16 { $0 }, [ $1 + 0 ];", "=c,l,b"(ptr addrspace(1) %13, i1 %10) #1, !dbg !15
  %15 = tail call i16 asm sideeffect "mov.u16 $0, 0x0;\0A\09@$2 ld.global.b16 { $0 }, [ $1 + 0 ];", "=c,l,b"(ptr addrspace(1) %13, i1 %10) #1, !dbg !15
  %16 = tail call i16 asm sideeffect "mov.u16 $0, 0x0;\0A\09@$2 ld.global.b16 { $0 }, [ $1 + 0 ];", "=c,l,b"(ptr addrspace(1) %13, i1 %10) #1, !dbg !15
  %17 = tail call i16 asm sideeffect "mov.u16 $0, 0x0;\0A\09@$2 ld.global.b16 { $0 }, [ $1 + 0 ];", "=c,l,b"(ptr addrspace(1) %13, i1 %10) #1, !dbg !15
  %18 = tail call i16 asm sideeffect "mov.u16 $0, 0x0;\0A\09@$2 ld.global.b16 { $0 }, [ $1 + 0 ];", "=c,l,b"(ptr addrspace(1) %13, i1 %10) #1, !dbg !15
  %19 = tail call i16 asm sideeffect "mov.u16 $0, 0x0;\0A\09@$2 ld.global.b16 { $0 }, [ $1 + 0 ];", "=c,l,b"(ptr addrspace(1) %13, i1 %10) #1, !dbg !15
  %20 = tail call i16 asm sideeffect "mov.u16 $0, 0x0;\0A\09@$2 ld.global.b16 { $0 }, [ $1 + 0 ];", "=c,l,b"(ptr addrspace(1) %13, i1 %10) #1, !dbg !15
  %21 = tail call i16 asm sideeffect "mov.u16 $0, 0x0;\0A\09@$2 ld.global.b16 { $0 }, [ $1 + 0 ];", "=c,l,b"(ptr addrspace(1) %13, i1 %10) #1, !dbg !15
  %22 = sext i32 %9 to i64, !dbg !16
  %23 = getelementptr half, ptr addrspace(1) %1, i64 %22, !dbg !16
  %24 = insertelement <2 x i16> undef, i16 %14, i64 0, !dbg !17
  %25 = insertelement <2 x i16> %24, i16 %15, i64 1, !dbg !17
  %26 = bitcast <2 x i16> %25 to i32, !dbg !17
  %27 = insertelement <2 x i16> undef, i16 %16, i64 0, !dbg !17
  %28 = insertelement <2 x i16> %27, i16 %17, i64 1, !dbg !17
  %29 = bitcast <2 x i16> %28 to i32, !dbg !17
  %30 = insertelement <2 x i16> undef, i16 %18, i64 0, !dbg !17
  %31 = insertelement <2 x i16> %30, i16 %19, i64 1, !dbg !17
  %32 = bitcast <2 x i16> %31 to i32, !dbg !17
  %33 = insertelement <2 x i16> undef, i16 %20, i64 0, !dbg !17
  %34 = insertelement <2 x i16> %33, i16 %21, i64 1, !dbg !17
  %35 = bitcast <2 x i16> %34 to i32, !dbg !17
  tail call void asm sideeffect "@$5 st.global.v4.b32 [ $4 + 0 ], { $0, $1, $2, $3 };", "r,r,r,r,l,b"(i32 %26, i32 %29, i32 %32, i32 %35, ptr addrspace(1) %23, i1 %10) #1, !dbg !17
  ret void, !dbg !18
}
```

### Benchmarking

This is the `jagged_to_padded_dense` kernel generated by TorchInductor. The kernel is converting a jagged tensor input (represented by separate `values` and `offsets` tensors) into the corresponding dense (padded) tensor `output`, using the scalar value `3.14` for the elements not present in the sparse jagged input representation:

```python
@triton.jit
def triton_kernel(values, offsets, output, batch_size, xnumel, XBLOCK: tl.constexpr):
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[:]
    xmask = xindex < xnumel

    x1 = xindex // 66560
    x2 = (xindex // 256) % 260
    x3 = xindex % 66560

    tmp0 = tl.load(offsets + (x1), xmask, eviction_policy="evict_last")
    tmp2 = tl.load(offsets + (1 + x1), xmask, eviction_policy="evict_last")
    tmp1 = tl.where(tmp0 < 0, tmp0 + 1 + batch_size, tmp0)
    tmp3 = tmp1 + x2
    tmp4 = tmp3 < tmp2
    tmp5 = tl.load(values + (x3 + (256 * tmp1)), tmp4 & xmask, other=0).to(tl.float32)
    tmp6 = tl.where(tmp4, tmp5, 3.14)

    tl.store(output + xindex, tmp6, xmask)
```

Below are the kernel runtimes before and after this PR as measured on different load factors of the jagged tensor input: i.e., the fraction of the data represented by the sparse jagged input (in other words, the average sequence length as a fraction of the maximum sequence length):

| load factor  | before this PR, ms | after this PR, ms | perf gain |
|---|---|---|---|
| 0.1 | 0.151 | 0.106 | 29% |
| 0.2 | 0.151 | 0.117 | 22% |
| 0.3 | 0.150 | 0.126 | 16% |
| 0.4 | 0.150 | 0.133 | 11% |
| 0.5 | 0.151 | 0.141 | 7% |
| 0.6 | 0.152 | 0.147 | 4% |
| 0.7 | 0.153 | 0.151 | 2% |
| 0.8 | 0.155 | 0.154 | 1% |
| 0.9 | 0.157 | 0.156 | 1% |
| 1.0 | 0.158 | 0.159 | 0% |

As it can be seen, on low load factors the heuristic introduced by this PR can lead to a performance gain of up to 29%. On high load factors, the memory I/O becomes the performance bottleneck making the index arithmetics deduplication less important.

### Test Plan

New `lit`-based unit tests have been added to `test/Conversion/dedup-by-constancy.mlir`:

```
$ lit test
-- Testing: 39 tests, 39 workers --
...

Testing Time: 0.70s
  Passed           : 38
  Expectedly Failed:  1
```

Jokeren commented on lib/Conversion/TritonGPUToLLVM/ElementwiseOpToLLVM.cpp:
```diff
@@ -512,8 +512,84 @@ class ElementwiseOpConversionBase
+
+    SmallVector<unsigned> elemsPerThread =
+        triton::gpu::getElemsPerThread(rtType);
+    int rank = elemsPerThread.size();
```
This isn't the actual `rank` of a tensor

Jokeren replied:
Or do you mean the heuristics do not apply when the number of elements is greater than the number of rank? Why?

aakhundov replied:
Would invoking `.size()` here not get the rank? My understanding was that `elemsPerThread.size()` and `constancy.size()` (as well as `rtType.getShape().size()` not used here) should all return the same number: the rank (the number of dimensions) of the tensor. Is this not true?

aakhundov replied:
Assuming that the rank (i.e., the `.size()`) of the `elemsPerThread` and `constancy` is the same, the downstream logic is to check that, in each of the dimensions in the rank, one of those is divisible by the other:

1. If the `elemsPerThread[i]` is divisible by the `constancy[i]`, we can partially deduplicate the values by replicating a single value in each `constancy[i]`-sized group `constancy[i] - 1` times.

2. If the `constancy[i]` is divisible by the `elemsPerThread[i]`, this means that, along the `i`th dimension, the constancy covers the whole bunch of elements in the thread and we can deduplicate them all by reusing the first value `elemsPerThread[i] - 1` times.

If neither of those conditions hold, we can't deduplicate, as there is no even coverage of `elemsPerThreads` by the `constancy` in one or more dimensions. Does the above make sense?

Jokeren replied:
> Would invoking .size() here not get the rank? My understanding was that elemsPerThread.size() and constancy.size() (as well as rtType.getShape().size() not used here) should all return the same number: the rank (the number of dimensions) of the tensor. Is this not true?

It's not true. `elemsPerThreads` could be more than the number of rank

Jokeren replied:
Oh, I guess I misunderstood the code. 

`elemsPerThreads.size()` should be the same as the rank. I read is as `elemsPerThreads`

Jokeren commented on lib/Conversion/TritonGPUToLLVM/ElementwiseOpToLLVM.cpp:
```diff
@@ -512,8 +512,84 @@ class ElementwiseOpConversionBase
+      // axis info (e.g., constancy) not available
+      return resultVals;
+    SmallVector<int64_t> constancy = axisInfo->getConstancy();
+    if (rank != constancy.size())
```
`constancy.size()` returns the rank of the tensor

aakhundov replied:
Indeed. Those checks are in the code to make sure that the downstream iteration over the constancy vector should work as expected.

Jokeren commented:
> @aakhundov This would benefit my kernels a lot as well. Related to this, would it be possible to achieve the same deduplication if you had non-contiguous repeating elements? For example, withx1 = xindex % 256 instead of x1 = xindex // 256.

I think the heuristic would be general since he checks the constancy of the result

Jokeren commented:
Do you have any benchmark that demonstrate the speedup? That would be great

aakhundov commented:
@francescocarzaniga

> For example, withx1 = xindex % 256 instead of x1 = xindex // 256.

In the case of `x1 = xindex % 256`, the resulting `x1` would have contiguity but not constancy. As the heuristic relies on the constancy for deduplication across the multiple (contiguous) elements in a single thread, I believe, dedup won't happen here.

francescocarzaniga commented:
@aakhundov I see, that's unfortunate. It would be great to have both kinds of deduplication working as they are complementary, but I'm no compiler expert so I have no idea how hard it would be to implement it.

Jokeren commented:
@aakhundov `python3 -m pre_commit run --all-files --verbose`

Jokeren commented:
> @aakhundov I see, that's unfortunate. It would be great to have both kinds of deduplication working as they are complementary, but I'm no compiler expert so I have no idea how hard it would be to implement it.

Contiguity wouldn't cause duplication I think, but `x % 256` does cause duplication if x is a vector and has constancy > 1 (e.g., [4, 4, 4, 4] % 256)?

francescocarzaniga commented:
> Contiguity wouldn't cause duplication I think, but `x % 256` does cause duplication if x is a vector and has constancy > 1 (e.g., [4, 4, 4, 4] % 256)?

Indeed in this case it would. What I am talking about is something like `[0, 1, 2, 3] % 2 -> [0, 1, 0, 1]` (this is complementary to `[0, 1, 2, 3] // 2 -> [0, 0, 1, 1]`). There is clearly some duplication going on here, but I assume it cannot be deduplicated in the same way.

Jokeren commented:
> Indeed in this case it would. What I am talking about is something like [0, 1, 2, 3] % 2 -> [0, 1, 0, 1] (this is complementary to [0, 1, 2, 3] // 2 -> [0, 0, 1, 1]). There is clearly some duplication going on here, but I assume it cannot be deduplicated in the same way.

There's no guarantee that [2,3] exist in the same thread as [0,1]. So in your case, maybe it is more of [0, 1, ... 512] % 256? So probably there's no duplication within the same thread?

There has been a long line of architecture research that automatically discovers duplication within and across threads and design function units to address the issues. The uniform register on nvidia GPUs is one way that addresses the duplicate values across threads within the same warp.
From the compiler perspective, I think we could gather more use cases and dive into the problems.

aakhundov commented:
> @aakhundov `python3 -m pre_commit run --all-files --verbose`

@Jokeren Fixed the linter. Could you approve the CI runs again? Thanks!

aakhundov commented:
> Do you have any benchmark that demonstrate the speedup? That would be great

Added the "Benchmark" section to the PR description.

aakhundov commented:
> Do you have any benchmark that demonstrate the speedup? That would be great

Added the "Benchmark" section to the PR description.

aakhundov commented:
@Jokeren apparently, it's generally unsafe to reuse LLVM `Value`s across different `sizePerThread`-sized blocks when `elemsPerThread` is larger. A few unit tests were failing due to this (e.g., `test_masked_load`), as some essential instructions got wrongly optimized out. I've added an additional condition for that and fixed the failing unit tests that I could verify on A100. Could you approve the CI again? Thanks!

Jokeren commented on lib/Conversion/TritonGPUToLLVM/ElementwiseOpToLLVM.cpp:
```diff
@@ -512,8 +512,104 @@ class ElementwiseOpConversionBase
+      return resultVals;
+    triton::gpu::BlockedEncodingAttr blockedEncoding =
+        encoding.dyn_cast<triton::gpu::BlockedEncodingAttr>();
+    if (!blockedEncoding)
```
Why doesn't it work for non-blocking encoding cases?

aakhundov replied:
With some trial and error, I've found out that if I don't constrain the encoding type to `BlcokedEncodingAttr` at this point, the downstream `triton::gpu::getElemsPerThread` crashes in the `test_fp8_dot_acc`. This is the error:

```
$ python3 -m pytest python/test/unit/language/test_core.py -k test_fp8_dot_acc
========================================================== test session starts ==========================================================
platform linux -- Python 3.11.5, pytest-7.4.2, pluggy-1.3.0
rootdir: /data/users/aakhundov/triton/python
collected 7000 items / 6992 deselected / 8 selected                                                                                     

python/test/unit/language/test_core.py Fatal Python error: Aborted

Current thread 0x00007f3f3defc480 (most recent call first):
  File "/data/users/aakhundov/triton/python/triton/compiler/compiler.py", line 167 in ttgir_to_llir
  File "/data/users/aakhundov/triton/python/triton/compiler/compiler.py", line 416 in <lambda>
  File "/data/users/aakhundov/triton/python/triton/compiler/compiler.py", line 517 in compile
  File "/data/users/aakhundov/triton/python/triton/runtime/jit.py", line 396 in launcher_body
  File "<string>", line 4 in matmul_kernel
  File "/data/users/aakhundov/triton/python/test/unit/language/test_core.py", line 3919 in test_fp8_dot_acc
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/_pytest/python.py", line 194 in pytest_pyfunc_call
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/pluggy/_callers.py", line 77 in _multicall
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/pluggy/_manager.py", line 115 in _hookexec
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/pluggy/_hooks.py", line 493 in __call__
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/_pytest/python.py", line 1792 in runtest
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/_pytest/runner.py", line 169 in pytest_runtest_call
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/pluggy/_callers.py", line 77 in _multicall
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/pluggy/_manager.py", line 115 in _hookexec
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/pluggy/_hooks.py", line 493 in __call__
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/_pytest/runner.py", line 262 in <lambda>
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/_pytest/runner.py", line 341 in from_call
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/_pytest/runner.py", line 261 in call_runtest_hook
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/_pytest/runner.py", line 222 in call_and_report
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/_pytest/runner.py", line 133 in runtestprotocol
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/_pytest/runner.py", line 114 in pytest_runtest_protocol
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/pluggy/_callers.py", line 77 in _multicall
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/pluggy/_manager.py", line 115 in _hookexec
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/pluggy/_hooks.py", line 493 in __call__
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/_pytest/main.py", line 350 in pytest_runtestloop
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/pluggy/_callers.py", line 77 in _multicall
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/pluggy/_manager.py", line 115 in _hookexec
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/pluggy/_hooks.py", line 493 in __call__
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/_pytest/main.py", line 325 in _main
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/_pytest/main.py", line 271 in wrap_session
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/_pytest/main.py", line 318 in pytest_cmdline_main
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/pluggy/_callers.py", line 77 in _multicall
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/pluggy/_manager.py", line 115 in _hookexec
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/pluggy/_hooks.py", line 493 in __call__
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/_pytest/config/__init__.py", line 169 in main
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/_pytest/config/__init__.py", line 192 in console_main
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/pytest/__main__.py", line 5 in <module>
  File "<frozen runpy>", line 88 in _run_code
  File "<frozen runpy>", line 198 in _run_module_as_main

Extension modules: numpy.core._multiarray_umath, numpy.core._multiarray_tests, numpy.linalg._umath_linalg, numpy.fft._pocketfft_internal, numpy.random._common, numpy.random.bit_generator, numpy.random._bounded_integers, numpy.random._mt19937, numpy.random.mtrand, numpy.random._philox, numpy.random._pcg64, numpy.random._sfc64, numpy.random._generator, torch._C, torch._C._fft, torch._C._linalg, torch._C._nested, torch._C._nn, torch._C._sparse, torch._C._special (total: 20)
Aborted (core dumped)
```

Also, a few lines below, I'm calling `blockedEncoding.getSizePerThread()` to fetch the size per thread. In principle, I could use `triton::gpu::getSizePerThread` instead, but just checked and can confirm that this leads to the same crash above in the `test_fp8_dot_acc`. If you have any tips on how to resolve the crash, I'm happy to lift the current constraint on the encoding type. Thanks!

Jokeren replied:
Probably because getSizePerThread returns wrong results

Jokeren replied:
Can you please leave a note in the code so we can review it later?

Jokeren replied:
Also, maybe it works fine already with slicedencoding?

aakhundov replied:
@Jokeren I've extended the coverage to `SliceEncodingAttr`, too. Also added a `TODO` comment for addressing the `test_fp8_dot_acc` crashing in future.

francescocarzaniga commented:
> There's no guarantee that [2,3] exist in the same thread as [0,1]. So in your case, maybe it is more of [0, 1, ... 512] % 256? So probably there's no duplication within the same thread?
> 
> There has been a long line of architecture research that automatically discovers duplication within and across threads and design function units to address the issues. The uniform register on nvidia GPUs is one way that addresses the duplicate values across threads within the same warp. From the compiler perspective, I think we could gather more use cases and dive into the problems.

My case is actually something like `[0, ..., 2^N] // M` (where M is not necessarily a power of 2), so there might be duplication in the same thread. I see what you mean however, that's why I was worried that optimising this case would not be as straightforward, but it would still be very useful.

ThomasRaoux commented:
so if I understand the optimization correctly the idea is that:
```
%10 = or i32 %9, 1, !dbg !11
%25 = sdiv i32 %10, 256, !dbg !13
```
can be changed to:
```
%25 = sdiv i32 %9, 256, !dbg !13
```

In general this is not correct for negative integers however this would be the case for unsigned integers (https://godbolt.org/z/7EqzzKEbM). Maybe the solution is to just cast this to an unsigned integer so that llvm will optimize things out. Technically llvm could also figure it out from the range but this would be more complicated.

aakhundov commented:
> so if I understand the optimization correctly the idea is that:
> 
> ```
> %10 = or i32 %9, 1, !dbg !11
> %25 = sdiv i32 %10, 256, !dbg !13
> ```
> 
> can be changed to:
> 
> ```
> %25 = sdiv i32 %9, 256, !dbg !13
> ```
> 
> In general this is not correct for negative integers however this would be the case for unsigned integers (https://godbolt.org/z/7EqzzKEbM). Maybe the solution is to just cast this to an unsigned integer so that llvm will optimize things out. Technically llvm could also figure it out from the range but this would be more complicated.

@ThomasRaoux I don't think that `or i32` is completely eliminated. From what I see in the example in the PR description, the following fragment with the 8 instances of each instruction in the original PTX:

```
  %9 = or i32 %8, %6, !dbg !11
  %10 = or i32 %9, 1, !dbg !11
  %11 = or i32 %9, 2, !dbg !11
  %12 = or i32 %9, 3, !dbg !11
  %13 = or i32 %9, 4, !dbg !11
  %14 = or i32 %9, 5, !dbg !11
  %15 = or i32 %9, 6, !dbg !11
  %16 = or i32 %9, 7, !dbg !11
  %17 = icmp slt i32 %9, %2, !dbg !12
  %18 = icmp slt i32 %10, %2, !dbg !12
  %19 = icmp slt i32 %11, %2, !dbg !12
  %20 = icmp slt i32 %12, %2, !dbg !12
  %21 = icmp slt i32 %13, %2, !dbg !12
  %22 = icmp slt i32 %14, %2, !dbg !12
  %23 = icmp slt i32 %15, %2, !dbg !12
  %24 = icmp slt i32 %16, %2, !dbg !12
  %25 = sdiv i32 %9, 256, !dbg !13
  %26 = sdiv i32 %10, 256, !dbg !13
  %27 = sdiv i32 %11, 256, !dbg !13
  %28 = sdiv i32 %12, 256, !dbg !13
  %29 = sdiv i32 %13, 256, !dbg !13
  %30 = sdiv i32 %14, 256, !dbg !13
  %31 = sdiv i32 %15, 256, !dbg !13
  %32 = sdiv i32 %16, 256, !dbg !13
  %33 = sext i32 %25 to i64, !dbg !14
  %34 = getelementptr half, ptr addrspace(1) %0, i64 %33, !dbg !14
  %35 = sext i32 %26 to i64, !dbg !14
  %36 = getelementptr half, ptr addrspace(1) %0, i64 %35, !dbg !14
  %37 = sext i32 %27 to i64, !dbg !14
  %38 = getelementptr half, ptr addrspace(1) %0, i64 %37, !dbg !14
  %39 = sext i32 %28 to i64, !dbg !14
  %40 = getelementptr half, ptr addrspace(1) %0, i64 %39, !dbg !14
  %41 = sext i32 %29 to i64, !dbg !14
  %42 = getelementptr half, ptr addrspace(1) %0, i64 %41, !dbg !14
  %43 = sext i32 %30 to i64, !dbg !14
  %44 = getelementptr half, ptr addrspace(1) %0, i64 %43, !dbg !14
  %45 = sext i32 %31 to i64, !dbg !14
  %46 = getelementptr half, ptr addrspace(1) %0, i64 %45, !dbg !14
  %47 = sext i32 %32 to i64, !dbg !14
  %48 = getelementptr half, ptr addrspace(1) %0, i64 %47, !dbg !14
```

gets deduplicated into a single instance of each instruction:

```
  %9 = or i32 %8, %6, !dbg !11
  %10 = icmp slt i32 %9, %2, !dbg !12
  %11 = sdiv i32 %9, 256, !dbg !13
  %12 = sext i32 %11 to i64, !dbg !14
  %13 = getelementptr half, ptr addrspace(1) %0, i64 %12, !dbg !14
```

as the result of each of those operations (or the downstream ones) has constancy covering the whole 8 values in the original PTX. Does this make sense?

ThomasRaoux commented:
> > so if I understand the optimization correctly the idea is that:
> > ```
> > %10 = or i32 %9, 1, !dbg !11
> > %25 = sdiv i32 %10, 256, !dbg !13
> > ```
> > 
> > 
> > can be changed to:
> > ```
> > %25 = sdiv i32 %9, 256, !dbg !13
> > ```
> > 
> > 
> > In general this is not correct for negative integers however this would be the case for unsigned integers (https://godbolt.org/z/7EqzzKEbM). Maybe the solution is to just cast this to an unsigned integer so that llvm will optimize things out. Technically llvm could also figure it out from the range but this would be more complicated.
> 
> @ThomasRaoux I don't think that `or i32` is completely eliminated. From what I see in the example in the PR description, the following fragment with the 8 instances of each instruction in the original PTX:
> 
> ```
>   %9 = or i32 %8, %6, !dbg !11
>   %10 = or i32 %9, 1, !dbg !11
>   %11 = or i32 %9, 2, !dbg !11
>   %12 = or i32 %9, 3, !dbg !11
>   %13 = or i32 %9, 4, !dbg !11
>   %14 = or i32 %9, 5, !dbg !11
>   %15 = or i32 %9, 6, !dbg !11
>   %16 = or i32 %9, 7, !dbg !11
>   %17 = icmp slt i32 %9, %2, !dbg !12
>   %18 = icmp slt i32 %10, %2, !dbg !12
>   %19 = icmp slt i32 %11, %2, !dbg !12
>   %20 = icmp slt i32 %12, %2, !dbg !12
>   %21 = icmp slt i32 %13, %2, !dbg !12
>   %22 = icmp slt i32 %14, %2, !dbg !12
>   %23 = icmp slt i32 %15, %2, !dbg !12
>   %24 = icmp slt i32 %16, %2, !dbg !12
>   %25 = sdiv i32 %9, 256, !dbg !13
>   %26 = sdiv i32 %10, 256, !dbg !13
>   %27 = sdiv i32 %11, 256, !dbg !13
>   %28 = sdiv i32 %12, 256, !dbg !13
>   %29 = sdiv i32 %13, 256, !dbg !13
>   %30 = sdiv i32 %14, 256, !dbg !13
>   %31 = sdiv i32 %15, 256, !dbg !13
>   %32 = sdiv i32 %16, 256, !dbg !13
>   %33 = sext i32 %25 to i64, !dbg !14
>   %34 = getelementptr half, ptr addrspace(1) %0, i64 %33, !dbg !14
>   %35 = sext i32 %26 to i64, !dbg !14
>   %36 = getelementptr half, ptr addrspace(1) %0, i64 %35, !dbg !14
>   %37 = sext i32 %27 to i64, !dbg !14
>   %38 = getelementptr half, ptr addrspace(1) %0, i64 %37, !dbg !14
>   %39 = sext i32 %28 to i64, !dbg !14
>   %40 = getelementptr half, ptr addrspace(1) %0, i64 %39, !dbg !14
>   %41 = sext i32 %29 to i64, !dbg !14
>   %42 = getelementptr half, ptr addrspace(1) %0, i64 %41, !dbg !14
>   %43 = sext i32 %30 to i64, !dbg !14
>   %44 = getelementptr half, ptr addrspace(1) %0, i64 %43, !dbg !14
>   %45 = sext i32 %31 to i64, !dbg !14
>   %46 = getelementptr half, ptr addrspace(1) %0, i64 %45, !dbg !14
>   %47 = sext i32 %32 to i64, !dbg !14
>   %48 = getelementptr half, ptr addrspace(1) %0, i64 %47, !dbg !14
> ```
> 
> gets deduplicated into a single instance of each instruction:
> 
> ```
>   %9 = or i32 %8, %6, !dbg !11
>   %10 = icmp slt i32 %9, %2, !dbg !12
>   %11 = sdiv i32 %9, 256, !dbg !13
>   %12 = sext i32 %11 to i64, !dbg !14
>   %13 = getelementptr half, ptr addrspace(1) %0, i64 %12, !dbg !14
> ```
> 
> as the result of each of those operations (or the downstream ones) has constancy covering the whole 8 values in the original PTX. Does this make sense?

right, this is because you use a sdiv, replacing that with a udiv should make the optimization happen.

Jokeren commented:
> I think you could just do xindex = xindex.to(tl.uint32) to force the index to be unsigned

I think that wouldn't deduplicate icmp and or

ThomasRaoux commented:
> > I think you could just do xindex = xindex.to(tl.uint32) to force the index to be unsigned
> 
> I think that wouldn't deduplicate icmp and or

right I missed the icmp part. How do we know that all the cmp will return the same value?

Jokeren commented:
> right I missed the icmp part. How do we know that all the cmp will return the same value?

It probably falls into this case and gets optimized by the PR?

```
        // Case 3: lhs all contiguous, rhs all constant
        // NOTE
        // lhs: 4 5 6 7
        // rhs: 4 4 4 4
```

aakhundov commented:
> > > I think you could just do xindex = xindex.to(tl.uint32) to force the index to be unsigned
> > 
> > 
> > I think that wouldn't deduplicate icmp and or
> 
> right I missed the icmp part. How do we know that all the cmp will return the same value?

Generally speaking, the code added by the PR checks the constancy property of the result of the lowered operator. And depending on what the constancy covers does the deduplication. In particular, the result of the `cmp` is deduplicated if the constancy associated with it (coming from the axis info analysis) allows it. Does this answer your question?

aakhundov commented:
@Jokeren @ThomasRaoux could you please approve the CI? Thanks!

Jokeren commented:
I'm not sure what's going on with the H100 node

aakhundov commented:
> I'm not sure what's going on with the H100 node

Perhaps we should try restarting the job?

Jokeren commented:
> > I'm not sure what's going on with the H100 node
> 
> Perhaps we should try restarting the job? Doesn't seem that the current one is going to commence.

It's showing waiting for resources. I've asked nvidia

aakhundov commented:
In case of rank-2+ results, with a non-default axis order (i.e., when the order is not `reversed(range(rank))`), the flat index mapping during deduplication was incorrect. I've added an axis reordering to fix the behavior in this case.

Jokeren commented on lib/Conversion/TritonGPUToLLVM/ElementwiseOpToLLVM.cpp:
```diff
@@ -598,6 +598,24 @@ class ElementwiseOpConversionBase
       // nothing to deduplicate
       return resultVals;
 
+    if (rank > 1) {
```
Can you use the `reorder` function in Utility.h?

aakhundov commented:
> > > I'm not sure what's going on with the H100 node
> > 
> > 
> > Perhaps we should try restarting the job? Doesn't seem that the current one is going to commence.
> 
> It's showing waiting for resources. I've asked nvidia

Seems that the H100 CI job is still pending. There may be something wrong with the runner.

ptillet commented:
Both our H100 CI an our backup are down for maintenance unfortunately. Bad timing... H100 CI will be back in a couple of days.

ptillet commented:
@aakhundov I merged main but I think the lit tests need to be adjusted based on https://github.com/openai/triton/pull/2526

aakhundov commented:
> @aakhundov I merged main but I think the lit tests need to be adjusted based on #2526

@ptillet Done. the lit tests are passing for me locally now. Could you please approve the CI? Thanks!

Jokeren commented:
Should be good to merge now

aakhundov commented:
@ptillet Do you think we can merge this? Thanks!
